src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -451,46 +451,33 @@ extern "C" JNIEXPORT int JVM_handle_linux_signal(int signo, siginfo_t* siginfo,
                                                int abort_if_unrecognized);
 #endif
 
-#if defined(AIX)
-
-// Set thread signal mask (for some reason on AIX sigthreadmask() seems
-// to be the thing to call; documentation is not terribly clear about whether
-// pthread_sigmask also works, and if it does, whether it does the same.
-bool set_thread_signal_mask(int how, const sigset_t* set, sigset_t* oset) {
-  const int rc = ::pthread_sigmask(how, set, oset);
-  // return value semantics differ slightly for error case:
-  // pthread_sigmask returns error number, sigthreadmask -1 and sets global errno
-  // (so, pthread_sigmask is more theadsafe for error handling)
-  // But success is always 0.
-  return rc == 0 ? true : false;
-}
-
-// Function to unblock all signals which are, according
-// to POSIX, typical program error signals. If they happen while being blocked,
-// they typically will bring down the process immediately.
-bool unblock_program_error_signals() {
+// Unblock all signals whose delivery cannot be deferred and which, if they happen
+//  while delivery is blocked, would cause crashes or hangs (JDK-8252533).
+void PosixSignals::unblock_error_signals() {
   sigset_t set;
   sigemptyset(&set);
   sigaddset(&set, SIGILL);
   sigaddset(&set, SIGBUS);
   sigaddset(&set, SIGFPE);
   sigaddset(&set, SIGSEGV);
-  return set_thread_signal_mask(SIG_UNBLOCK, &set, NULL);
-}
-
+  // We also unblock SIGTRAP, which is not in the list of POSIX lists of signals
+  //  which would cause undefined behavior when blocked
+  //  (https://pubs.opengroup.org/onlinepubs/009695399/functions/sigprocmask.html)
+  //  but in our experience cause the same problems when occurring when blocked.
+  sigaddset(&set, SIGTRAP);
+  int rc = ::pthread_sigmask(SIG_UNBLOCK, &set, NULL);
+#ifdef ASSERT
+  if (rc != 0) {
+    log_warning(os)("pthread_sigmask failed (%d)", errno);
```

Comment:
pthread_sigmask doesn't set errno, it returns the error code directly. But as there is only one defined error code for pthread_sigmask (EINVAL) this warning seem unnecessary. The usual assert_status would suffice.

Reply:
Good catch.

I thought about assert too. assert is not a good idea, since we would trigger our assertion poison page, and since segv may still be blocked, this would cause exactly the problem I want to prevent.

But since the only error is EINVAL, and we know what we feed into this function, I'll just remove the assert.